pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -676,6 +693,17 @@ func (adc *attachDetachController) syncPVCByKey(logger klog.Logger, key string)
 func (adc *attachDetachController) processVolumesInUse(
 	logger klog.Logger, nodeName types.NodeName, volumesInUse []v1.UniqueVolumeName) {
 	logger.V(4).Info("processVolumesInUse for node", "node", klog.KRef("", string(nodeName)))
+	// determine if VolumesInUse changes exists
+	adc.cacheVolumesInUseLock.RLock()
+	cachedVolumes, ok := adc.cacheVolumesInUse[nodeName]
+	if ok && time.Now().Sub(cachedVolumes.lastUpdate) < CacheVolumesInUseExpire {
+		if cachedVolumes.volumes.Len() == len(volumesInUse) && cachedVolumes.volumes.HasAll(volumesInUse...) {
+			logger.V(4).Info("processVolumesInUse skipped for node", "node", klog.KRef("", string(nodeName)))
+			return
+		}
+	}
+	adc.cacheVolumesInUseLock.RUnlock()
+
 	for _, attachedVolume := range adc.actualStateOfWorld.GetAttachedVolumesForNode(nodeName) {
```

Comment:
I wonder, since you're creating a set of volumes in use anyway, if you could do a set intersection with the attached volumes and save the n^2 loop here.

It may not matter too much since the number of volumes on any one node is small.

Reply:
Changed to n^1 loop done.